# Use version control (Git), and create a script to automatically visualize my favorite database
Author: Nikolaj Pagh Kristensen

Background: My two favorite databases that I often use in my project contains information T-cell targets and T-cell receptor(TCR) sequences. These are: IEDB(http://www.iedb.org/) and VDJdb (https://vdjdb.cdr3.net/ named after the V, D and J segments of the TCR). I wish to create a script that automatically imports a subset of the database, calculate summary statistics for my favorite groups of T-cells. Furthermore, I wish to visualize summary statistics of the database in it's current form.

To break it up into subgoals:
1. Create a github repository",
1. Create the necessary initial files according to public guides (https://kbroman.org/github_tutorial/pages/init.html)
1. Look for an API to download IEDB data with (if possible). Download only data relevant for CD8 T cells. 
1. What epitope length is cataloged for each HLA?
1. What virual antigens are frequently targeted by T-cells?
1. What viruses do the database contain information on, and how many epitopes?

Extra things:
1. Do something similar but with VDJdb
2. Perform sequence analysis of epitopes. Generate clusters of epitopes based on sequence similarity.

# Start a new github repository and start keeping track of version tags
From the terminal I will execute the following,
1. Make new dirs (Python_projects and Python_projects/Course_work),
2. Initialize git by typing git init within the Python_course_report folder:`git init`,
3. Generate a README file (Markdown) as well as a .gitingore `touch .gitignore`,
1. Use `git add` to add files to the staging area
1. Use `git commit -m` to commint the stagining area together with a required comment

I double cheched my commit by using `git log`

Version numbering is a good practice: https://stackoverflow.com/questions/37814286/how-to-manage-the-version-number-in-git
    The general nomenclature is: `[major].[minor].[patch]-[build/beta/rc]`

1. Use initial `git tag -a v0.1.0 -m "Initial version"`
2. inspect tags using `git describe` succesfully returned a list of versions for the git repo


# Scrape IEDB for data
IEDB released an API to collect information directly on the command line or in your Jupyter notbook. The API postREST platform (https://postgrest.org/en/stable/), which has a list of functions and operators that are useful to know (https://postgrest.org/en/v7.0.0/api.html. For more information how it was implemented in the IEDB API see: https://query-api.iedb.org/docs/swagger/ and seek help here: https://help.iedb.org/hc/en-us/articles/4402872882189-Immune-Epitope-Database-Query-API-IQ-API-

## Core endpoints from the API
While there are many endpoints provided, we expect the majority of users will want to search against one or more of the following tables, which correspond to the tabbed search results on the IEDB. 

    epitope_search
    antigen_search
    tcell_search (assays) 
    bcell_search (assays)
    mhc_search (assays)
    tcr_search (receptors)
    bcr_search (receptors)
    reference_search

We're mainly interested in:

    epitope_search
    tcell_search (assays) 
    tcr_search (receptors)
   
A good example is found in file: use_case_1a.ipynb. As this is not my example, I have chosen to hide it with .gitignore
***
## Downloading information on a single epitope in a tabular format
This is described in the tutorial: use_case_1a.ipynb. 
<br>Let's start with the necessary packages:

In [2]:
import requests
import json
import pandas as pd
from io import StringIO

Let's define the url that we will be sending requests to. 

In [3]:
base_uri='https://query-api.iedb.org'

The user-case recommends a function to print the CURL command given a request:

In [22]:
# funciton to print the CURL command given a request
    #req is the request, a request
def print_curl_cmd(req):
    url = req.url
    print("This is the request url used with the curl cmd:") #My addition
    print("curl -X 'GET' '" + url + "'")

The first search should be *simple*, meaning that we will download all information related to a single epitope (my favorite epitope)
1. Define my_favorite_epitope
1. Define my search paramters (Fx: search_params={ 'linear_sequence': 'eq.SIINFEKL'}
2. Define the table_name (what information to search for
3. Define the full url of the table that we search for
4. fetch the result by parsing the full URL to requests.get, searching for your search paramters, saving the download in results. 
5. print the curl -x get command 

In [28]:
my_favorite_epitope = "HPVGEADYFEY"  #HPVGEADYFEY is one of the longest, dominant epitopes on B*3501
search_params={ 'linear_sequence': 'eq.'+my_favorite_epitope} #eq. here is a special postREST operator it abbreviates "equal"
table_name='epitope_search'
full_url=base_uri + '/' + table_name
result = requests.get(full_url, params=search_params)
print_curl_cmd(result)

This is the request url used with the curl cmd:
curl -X 'GET' 'https://query-api.iedb.org/epitope_search?linear_sequence=eq.HPVGEADYFEY'


We have the results. Inspect it:

    what type is it?

Object `request.get` not found.


In [29]:
print("The downloaded file is of type", type(result))

The downloaded file is of type <class 'requests.models.Response'>


We need to transform *class: 'requests.models.Response'* into something useful. The IEDB API Tutorial recommends turning into a data frame right away

In [30]:
df = pd.json_normalize(result.json())
print("This is a data from of the following type:", type(df))
df

This is a data from of the following type: <class 'pandas.core.frame.DataFrame'>


structure_id       structure_iri structure_descriptions  \
0         24536  IEDB_EPITOPE:24536          [HPVGEADYFEY]   

                             curated_source_antigens  structure_type  \
0  [{'accession': 'AFY97913.1', 'name': 'EBNA-1',...  Linear peptide   

  linear_sequence e_modification  linear_sequence_length  \
0     HPVGEADYFEY           None                      11   

                                      iedb_assay_ids  \
0  [5452, 1328890, 1328891, 1382633, 1382634, 138...   

                                     iedb_assay_iris  ... bcell_ids  \
0  [IEDB_ASSAY:1328890, IEDB_ASSAY:1328891, IEDB_...  ...      None   

  bcell_iris                                        elution_ids  \
0       None  [1383248, 1383263, 1403131, 1403132, 1450643, ...   

                                        elution_iris  \
0  [IEDB_ASSAY:1383248, IEDB_ASSAY:1383263, IEDB_...   

            reference_types  \
0  [Literature, Submission]   

                                          pubmed_ids  \
0  [11120837, 12576337, 14694109, 15148339, 15148...   

                                 reference_abstracts  \
0  [A classic feature of antigen presentation for...   

                                    reference_titles  \
0  [Allelic polymorphism in the T cell receptor a...   

                                   reference_authors  \
0  [Barbara Savoldo; John A Goss; Markus M Hammer...   

                                     reference_dates  
0  [2000, 2003, 2004, 2005, 2006, 2007, 2008, 200...  

[1 rows x 83 columns]

The current dataframe has a single observation and 83 columns. That is because each epitope is represented as a row when using "epitope_search" with 83 different features.

Next aims:
- Inspect all column names
- Let's redo the epitope search, but let's limit the number of columns to receive. 
- Let's performn the epitope search for a given HLA allele instead (e.g. B\*35:01)

In [31]:
#Inspecting all column names
df.columns

Index(['structure_id', 'structure_iri', 'structure_descriptions',
       'curated_source_antigens', 'structure_type', 'linear_sequence',
       'e_modification', 'linear_sequence_length', 'iedb_assay_ids',
       'iedb_assay_iris', 'reference_ids', 'reference_iris', 'submission_ids',
       'submission_iris', 'pdb_ids', 'chebi_ids', 'qualitative_measures',
       'mhc_allele_evidences', 'antibody_isotypes', 'direct_ex_vivo_bool',
       'receptor_ids', 'receptor_group_ids', 'tcr_receptor_group_ids',
       'bcr_receptor_group_ids', 'receptor_group_iris',
       'tcr_receptor_group_iris', 'bcr_receptor_group_iris', 'receptor_types',
       'receptor_names', 'receptor_chain1_types', 'receptor_chain2_types',
       'receptor_chain1_full_seqs', 'receptor_chain2_full_seqs',
       'receptor_chain1_cdr1_seqs', 'receptor_chain2_cdr1_seqs',
       'receptor_chain1_cdr2_seqs', 'receptor_chain2_cdr2_seqs',
       'receptor_chain1_cdr3_seqs', 'receptor_chain2_cdr3_seqs',
       'host_organism_iri

These are the column names in Epitope_search: 
          
Of interest are:
- curated_source_antigens
- linear_sequence
- e_modification
- linear_sequence_length
- mhc_allele_names
- source_organism_names
- iedb_assay_ids
- reference_ids

**These also doubles as search paramters**
<br>See: https://query-api.iedb.org/docs/swagger/#/epitope_search/get_epitope_search


In [35]:
#This code now selects the the exact columns we are interested in from within the IEDB API before the resulting df is generated
search_params={  'linear_sequence': 'eq.'+my_favorite_epitope,
                'select': 'curated_source_antigens, linear_sequence, e_modification, linear_sequence_length, mhc_allele_names,\
                source_organism_names, iedb_assay_ids, reference_ids'}
result = requests.get(full_url, params=search_params)
print_curl_cmd(result)
df = pd.json_normalize(result.json())
df

This is the request url used with the curl cmd:
curl -X 'GET' 'https://query-api.iedb.org/epitope_search?linear_sequence=eq.HPVGEADYFEY&select=curated_source_antigens%2C+linear_sequence%2C+e_modification%2C+linear_sequence_length%2C+mhc_allele_names%2C++++++++++++++++source_organism_names%2C+iedb_assay_ids%2C+reference_ids'


curated_source_antigens linear_sequence  \
0  [{'accession': 'AFY97913.1', 'name': 'EBNA-1',...     HPVGEADYFEY   

  e_modification  linear_sequence_length  \
0           None                      11   

                                    mhc_allele_names  \
0  [HLA-A*02:01, HLA-A2, HLA-B*07:02, HLA-B*08:01...   

                               source_organism_names  \
0  [Human herpesvirus 4 (Epstein Barr virus), Hum...   

                                      iedb_assay_ids  \
0  [5452, 1328890, 1328891, 1382633, 1382634, 138...   

                                       reference_ids  
0  [589, 1002081, 1004304, 1004493, 1004495, 1004...

From the tutorial: 
    "Note the additional complexity in the URL of the last query **There are two parameters (linear_sequence & select)**, multiple values for the latter parameter, and many URL escape codes for the commas. Python's 'request' module handles this all for you, but one should be aware that all portions of the query need to be URL-escaped."

# Downloading information on a all epitopes in Epitope_search

In [36]:
#See title
search_params={  'select': 'curated_source_antigens, linear_sequence, e_modification, linear_sequence_length, mhc_allele_names,\
                source_organism_names, iedb_assay_ids, reference_ids'} 
table_name='epitope_search'
full_url=base_uri + '/' + table_name
result = requests.get(full_url, params=search_params)
print_curl_cmd(result)
df = pd.json_normalize(result.json())

This is the request url used with the curl cmd:
curl -X 'GET' 'https://query-api.iedb.org/epitope_search?select=curated_source_antigens%2C+linear_sequence%2C+e_modification%2C+linear_sequence_length%2C+mhc_allele_names%2C++++++++++++++++source_organism_names%2C+iedb_assay_ids%2C+reference_ids'


curated_source_antigens       linear_sequence  \
0     [{'accession': 'P26664.3', 'name': 'Genome pol...             GDLCGSVFL   
1     [{'accession': 'AAB59097.1', 'name': 'exotoxin...  GDLDPSSIPDKEQAISALPD   
2     [{'accession': 'P08318.1', 'name': 'Large stru...       GDLFSGDEDSDSSDG   
3     [{'accession': 'YP_001623301.1', 'name': 'cell...       GDLGKKGFEDGDLVV   
4     [{'accession': 'P03306.2', 'name': 'Genome pol...               GDLGSIA   
...                                                 ...                   ...   
9995  [{'accession': 'P03468.2', 'name': 'Neuraminid...            ITYKNSTWVK   
9996  [{'accession': 'AAB00381.1', 'name': 'S1 glyco...  ITYKVMREVRALAYFVNGTA   
9997  [{'accession': 'AAA69398.1', 'name': 'D7R', 'i...             ITYLMNRFK   
9998  [{'accession': 'AAB96556.1', 'name': '68k anky...         ITYLMNRFKNIDI   
9999  [{'accession': 'P13811.1', 'name': 'Heat-labil...            ITYLTETKID   

     e_modification  linear_sequence_length  \
0              None                       9   
1              None                      20   
2              None                      15   
3              None                      15   
4              None                       7   
...             ...                     ...   
9995           None                      10   
9996           None                      20   
9997           None                       9   
9998           None                      13   
9999           None                      10   

                                       mhc_allele_names  \
0                               [H2-Kk, Mamu-A1*011:01]   
1                                                  None   
2                                                  None   
3                                                  None   
4                                                  None   
...                                                 ...   
9995                                            [H2-Db]   
9996                                          [chicken]   
9997  [HLA-A*03:01, HLA-A*11:01, HLA-A*30:01, HLA-A*...   
9998                                            [human]   
9999                                    [H2-d class II]   

                                  source_organism_names  \
0                       [Hepatitis C virus (isolate 1)]   
1                              [Pseudomonas aeruginosa]   
2         [Human herpesvirus 5 (Human cytomegalovirus)]   
3                          [Renibacterium salmoninarum]   
4        [Foot-and-mouth disease virus (strain A10-61)]   
...                                                 ...   
9995  [Influenza A virus (A/Puerto Rico/8/1934(H1N1)...   
9996  [Avian infectious bronchitis virus (strain Vic...   
9997  [Camelpox virus M-96, Vaccinia virus Copenhage...   
9998       [Modified Vaccinia Ankara virus (MVA virus)]   
9999  [Escherichia coli ETEC H10407 (Escherichia col...   

                                         iedb_assay_ids  \
0                                    [1182763, 1182764]   
1                                             [1269457]   
2                                             [1382552]   
3                                    [1478738, 1478739]   
4     [1480902, 1480903, 1480906, 1480907, 1480909, ...   
...                                                 ...   
9995                                 [1006289, 1006290]   
9996      [1498067, 1498075, 1498257, 1498284, 1500789]   
9997  [1301052, 1301053, 1301054, 1301055, 1301056, ...   
9998                                          [1502511]   
9999  [1269927, 1270260, 1270540, 1270550, 1273658, ...   

                                          reference_ids  
0                                             [1000983]  
1                                             [1001199]  
2                                             [1004651]  
3                                             [1006177]  
4                                    [1008339, 1010423]  
...                

Inspect the data frame: 
- shape
- sample
- linear_sequence_length

In [37]:
df.shape

(10000, 8)

**HUH**, it appear wierd that there are exactly 10.000 rows. There should be more... 
How many epitopes is within IEDB currently?
- (26-08-2021) 39868 epitopes are available. There most be a limit or something on the API.. 

Interesting: 
    The default request is 10.000, any more will be hidden on a different "page"
    <br> *By default, the IQ-API has a maximum page size of 10,000 records. In practice, this means that queries that result in more than 10,000 results will be divided into pages and only the first 10,000 records will be returned by the initial query. The API will always return a count of the records matching the query, as well as the number of pages of results. 
    
There ARE a way to download all hits. But, it requires me to use the curl command in the terminal.

## Download Epitope_search and Assay_search by using the terminal

In [50]:
#Define the epitope_search url
search_params={  'select': 'curated_source_antigens, linear_sequence, e_modification, linear_sequence_length, mhc_allele_names,\
                source_organism_names, iedb_assay_ids, reference_ids'} 
table_name='epitope_search'
print(base_uri)
full_url=base_uri + '/' + table_name
print(full_url)

https://query-api.iedb.org
https://query-api.iedb.org/epitope_search


In [51]:
#Define the Antigen_search url
search_params={  'select': 'curated_source_antigens, linear_sequence, e_modification, linear_sequence_length, mhc_allele_names,\
                source_organism_names, iedb_assay_ids, reference_ids'} 
table_name='antigen_search'
print(base_uri)
full_url=base_uri + '/' + table_name
print(full_url)

https://query-api.iedb.org
https://query-api.iedb.org/antigen_search


#In the terminal: 
```bash
$ curl -I "https://query-api.iedb.org/epitope_search" -H 'Prefer: count=exact'
HTTP/1.1 206 Partial Content
Server: nginx/1.14.2
Date: Thu, 26 Aug 2021 17:33:11 GMT
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Content-Range: 0-9999/1535288
Content-Location: /epitope_search
```

From the abovee, I can see that:
- There are 1535288 entries in the epitope_search database

I can access any part of the database above 9999 by specifying the offset paramter in the url

```bash
$ curl -I "https://query-api.iedb.org/antigen_search?offset=10000" -H 'Prefer: count=exact'
HTTP/1.1 206 Partial Content
Server: nginx/1.14.2
Date: Thu, 26 Aug 2021 17:41:01 GMT
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Content-Range: 10000-19999/73218
Content-Location: /antigen_search?offset=10000
```


### Downloading information has to be done in chunks followed with merging

Aims:
- Create a simple bash script to download, store and bind datasets together for Epitope_search
- Do the same for assay_search